### PR TITLE
Column and existence verification for score assets

### PR DESCRIPTION
**Pull request by juliexxia**
On load and on asset picker change, verify the feature collection specified to go into the score asset (1) exists and (2) has the column headers we expect for that type of asset. 

This moves us more towards being very locked down on census column headers. 

#279

### CODE REVIEW

File: docs/import/manage_disaster.js [hunk 1]
```diff
@@ -512,6 +535,97 @@ function createAssetDropdown(
...
+/**
+ * Checks the given asset for the given columns and prints a result message.
+ * @param {string} asset
+ * @param {string} type values from the first index of each entry in {@code
+ *     scoreAssetTypes}
+ * @param {string} state e.g. 'WA'
+ * @param {Array<string>} expectedColumns
+ * @return {?Promise<unknown>} returns null if column checking wasn't needed.
+ * Otherwise returns a promise that resolves when column checking is finished
+ * and select border color is updates.
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: nit: "is updated"?

File: docs/import/manage_disaster.js [hunk 2]
```diff
@@ -404,22 +417,32 @@ function initializeScoreSelectors(states) {
...
             createAssetDropdown(stateAssets.get(state), statePropertyPath)
-                .prop('id', 'select-' + id + '-' + state),
-            statePropertyPath)));
+                .prop('id', 'select-' + id + '-' + state)
+                .on('change',
+                    (event) => onNonDamageAssetSelect(
+                        event, statePropertyPath, expectedColumns, type, state))
+                .prop('style', 'border:2px');
+        row.append(
+            createTd().append($(document.createElement('span'))
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: Why can't you add hover text to the select directly? When I give it a title attribute, it seems to hover fine.
hover-screenshot

File: cypress/integration/unit_tests/manage_disaster_test.js [hunk 3]
```diff
@@ -313,6 +320,68 @@ describe('Unit tests for manage_disaster.js', () => {
                 ' and must specify either damage asset or map bounds');
   });
 
+  it.only('does column verification', () => {
```
[the remark this answers is not in the document]
Comment (janakdr) [reply cypress/integration/unit_tests/manage_disaster_test.js]: I think you can do this pretty easily (and change the other test cases to explicitly know when the promise finishes) by wrapping convertEeObjectToPromise and using a CallbackLatch. That will give you full control and be pretty robust.

[the remark this answers is not in the document]
Comment (janakdr) [reply cypress/integration/unit_tests/manage_disaster_test.js]: I think you can do this pretty easily (and change the other test cases to explicitly know when the promise finishes) by wrapping convertEeObjectToPromise and using a CallbackLatch. That will give you full control and be pretty robust.

[the remark this answers is not in the document]
Comment (juliexxia) [reply cypress/integration/unit_tests/manage_disaster_test.js]: Still figuring out this test, FYI. Wanted to send the rest of the updates back.

File: cypress/integration/unit_tests/manage_disaster_test.js [hunk 4]
```diff
@@ -313,6 +320,115 @@ describe('Unit tests for manage_disaster.js', () => {
...
+        '#select-asset-selection-row-buildings-NY', 'rgb(0, 128, 0)');
+    checkHoverText(
+        '#select-asset-selection-row-buildings-NY', 'No expected columns');
+    setFirstSelectInScoreRowTo(4, 'None');
+    checkSelectBorder(
+        '#select-asset-selection-row-buildings-NY', 'rgb(255, 255, 255)');
+    checkHoverText('#select-asset-selection-row-buildings-NY', '');
+  });
+
+  it('tries to set set a missing asset', () => {
```
Comment (janakdr) [inline cypress/integration/unit_tests/manage_disaster_test.js]: nit: "set set"

File: docs/import/manage_disaster.js [hunk 5]
```diff
@@ -404,22 +416,29 @@ function initializeScoreSelectors(states) {
...
+    const id = assetSelectionRowPrefix + type;
     const propertyPath = scoreAssetType[1];
+    const expectedColumns = scoreAssetType[3];
     const row = $('#' + id);
     removeAllButFirstFromRow(row);
     for (const state of states) {
       if (stateAssets.get(state)) {
         const statePropertyPath = propertyPath.concat([state]);
-        row.append(createTd().append(addAssetDataChangeHandler(
+        const select =
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: nit: can leave this as a variable for readability, but could also inline now

File: docs/import/manage_disaster.js [hunk 6]
```diff
@@ -512,6 +530,104 @@ function createAssetDropdown(
...
+ * @param {string} type values from the first index of each entry in {@code
+ *     scoreAssetTypes}
+ * @param {string} state e.g. 'WA'
+ * @param {Array<string>} expectedColumns
+ * @return {Promise<void>} returns null if column checking wasn't needed.
+ * Otherwise returns a promise that resolves when column checking is finished
+ * and select border color is updated.
+ */
+function verifyAsset(asset, type, state, expectedColumns) {
+  // TODO: disable kick off button until all green?
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: I would think not? Maybe we could put a warning in the kick-off button text to address the red parts?

File: docs/import/manage_disaster.js [hunk 7]
```diff
@@ -512,6 +530,104 @@ function createAssetDropdown(
...
+  const tdId = type + '-' + state;
+  const select = $('#select-' + assetSelectionRowPrefix + type + '-' + state);
+  lastSelectedAsset.set(tdId, asset);
+  const assetMissingErrorFunction = (err) => {
+    if (err.includes('\'' + asset + '\' not found.')) {
+      updateColorAndHover(select, 'red', 'Error! asset could not be found.');
+    } else {
+      console.error(err);
+      updateColorAndHover(select, 'red', 'Unknown error.');
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: Any reason not to show the error in the hover message? Like, "Unknown error: ..."

File: docs/import/manage_disaster.js [hunk 8]
```diff
@@ -512,6 +535,97 @@ function createAssetDropdown(
...
+  const span = $('#' + tdId + '-hover');
+  const select = span.children('select');
+  lastSelectedAsset.set(tdId, asset);
+  if (asset === '') {
+    return updateColorAndHover(select, 'white', span, '');
+  } else if (expectedColumns.length === 0) {
+    return updateColorAndHover(select, 'green', span, 'No expected columns');
+  } else {
+    updateColorAndHover(select, 'yellow', span, 'Checking columns...');
+    return convertEeObjectToPromise(getColumnsStatus(asset, expectedColumns))
```
[the remark this answers is not in the document]
Comment (janakdr) [reply docs/import/manage_disaster.js]: Yeah, I think it's fine, as we discussed we don't want to zero out user choices just from page load.

File: docs/import/manage_disaster.js [hunk 9]
```diff
@@ -512,6 +530,104 @@ function createAssetDropdown(
...
+                  select, 'green', 'Success! asset has all expected columns');
+            }
+          }
+        })
+        .catch(assetMissingErrorFunction);
+  }
+}
+
+/**
+ * Updates the border of the select and the hover text of the span.
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: nit: s/span/select/

File: docs/import/manage_disaster.js [hunk 10]
```diff
@@ -512,6 +530,104 @@ function createAssetDropdown(
...
+ * @param {JQuery<HTMLSelectElement>} select
+ * @param {string} color
+ * @param {string} title
+ */
+function updateColorAndHover(select, color, title) {
+  select.prop('style', 'border:2px solid ' + color).prop('title', title);
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: nit: can you just specify `border: <color>` here and set `border: 2px solid` elsewhere (like, even, in a manage_disaster.css class)? Don't actually know if CSS works that way

File: docs/import/manage_disaster.js [hunk 11]
```diff
@@ -559,57 +553,65 @@ function onNonDamageAssetSelect(
...
 /**
- * Checks the given asset for the given columns and prints a result message.
+ * Verifies an asset exists and has the expected columns.
  * @param {string} asset
  * @param {string} type values from the first index of each entry in {@code
  *     scoreAssetTypes}
  * @param {string} state e.g. 'WA'
  * @param {Array<string>} expectedColumns
- * @return {?Promise<unknown>} returns null if column checking wasn't needed.
+ * @return {Promise<void>} returns null if column checking wasn't needed.
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: do we ever return null now? Might also want to rephrase to mention existence checking.

File: docs/import/manage_disaster.js (hunk 7 above)
Comment (juliexxia) [reply docs/import/manage_disaster.js]: Ah, sure.

File: docs/import/manage_disaster.js (hunk 6 above)
Comment (juliexxia) [reply docs/import/manage_disaster.js]: Updated, we can chat later

File: docs/import/manage_disaster.js (hunk 11 above)
Comment (juliexxia) [reply docs/import/manage_disaster.js]: Still do when we don't have an asset to check (set to None)

File: docs/import/manage_disaster.js (hunk 5 above)
Comment (juliexxia) [reply docs/import/manage_disaster.js]: I think helpful for readability so going to leave.

File: docs/import/manage_disaster.js [hunk 12]
```diff
@@ -512,6 +531,104 @@ function createAssetDropdown(
...
+
+/**
+ * Verifies an asset exists and has the expected columns.
+ * @param {string} asset
+ * @param {string} type values from the first index of each entry in {@code
+ *     scoreAssetTypes}
+ * @param {string} state e.g. 'WA'
+ * @param {Array<string>} expectedColumns
+ * @return {Promise<void>} returns null if there was no asset to check.
+ * Otherwise returns a promise that resolves when existence and column checking
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: super-nit: I think should be indented?

File: docs/import/manage_disaster.js [hunk 13]
```diff
@@ -404,22 +417,29 @@ function initializeScoreSelectors(states) {
...
-        row.append(createTd().append(addAssetDataChangeHandler(
+        const select =
             createAssetDropdown(stateAssets.get(state), statePropertyPath)
-                .prop('id', 'select-' + id + '-' + state),
-            statePropertyPath)));
+                .prop('id', 'select-' + id + '-' + state)
+                .on('change',
+                    (event) => onNonDamageAssetSelect(
+                        event, statePropertyPath, expectedColumns, type, state))
+                .prop('style', 'border:2px solid');
```
Comment (janakdr) [inline docs/import/manage_disaster.js]: Can we put this in a css file and just give this select a class? If you don't want to do that, can we not use the .css() function here?

File: cypress/integration/unit_tests/manage_disaster_test.js [hunk 14]
```diff
@@ -718,3 +881,63 @@ function getFirstTdInScoreRow(rowNum) {
...
+/**
+ * Returns a latch that controls the logic of {@link convertEeObjectToPromise}.
+ * @return {CallbackLatch}
+ */
+function getConvertEeObjectToPromiseLatch() {
+  const latch = new CallbackLatch();
+  const oldConvert = MapUtil.convertEeObjectToPromise;
+  MapUtil.convertEeObjectToPromise = (eeObject) => {
+    MapUtil.convertEeObjectToPromise = oldConvert;
+    return latch.delayedCallback(() => convertEeObjectToPromise(eeObject))();
```
[the remark this answers is not in the document]
Comment (juliexxia) [reply cypress/integration/unit_tests/manage_disaster_test.js]: Think I got it but still wrapping my head around the async/await situation. Is this better than Promise.all([MapUtil.convertEeObjectToPromise(eeObject), promise]).then(...) ?

**Pull request merged**